### ISSUE TITLE
Support adding to the Deployment DependsOn list

### ISSUE THREAD

**Issue by glb** [serverless/serverless#7338]
# Feature Proposal

## Description

I have an AWS::ApiGateway::Method that I think I need to specify manually in the `Resources` section. There doesn't seem to be a way to add this resource to the `DependsOn` list for the generated `AWS::ApiGateway::Deployment` ... because the ID for the deployment is generated, I can't put something in `Resources` and have it merged the way I can with other stack resources.

Because I can't make the deployment depend on the resource creation, the resource might not be created by the time the deployment starts, which means that the resource won't be available in my deployed API.

Proposal: allow specifying `deployment.dependsOn: [resourceId...]`.

Sample configuration:

```yaml
provider:
  apigateway:
    deployment:
      dependsOn:
        - resourceID1
        - resourceID2
```

Related code: https://github.com/serverless/serverless/blob/3910df1ba6a8b39367ce8d51adb90216251be2ba/lib/plugins/aws/package/compile/events/apiGateway/lib/deployment.js#L20

**Comment by medikoo** [comment]
Thanks @glb for report.

Probably best solution, would be to come up with something more generic that will help address other similar cases, and also allow us to keep CloudFormation template extensions in one place (I wouldn't propagate it to `provider` namespace, as we don't keep there CF logic)

Currently we allow to extend CF resources via `resources`. So maybe we can introduce a way to address  framework generated resources with not predictable id's, via alias, as e.g. below:

```yaml
resources:
  Resources:
    $ApiGatewayDeployment:
      DependsOn:
        - resourceID1
```

What do you think?

---

Also your specific issue is solved by [serverless-plugin-bind-deployment-id](https://www.npmjs.com/package/serverless-plugin-bind-deployment-id), so you may rely on that until we have that solved in a framework

**Comment by glb** [comment]
Thanks @medikoo ... unfortunately, using the mechanism suggested by `serverless-plugin-bind-deployment-id` results in the wrong output from the merge.

Before:

```yml
    "ApiGatewayDeploymentXXXXXXXXX": {
      "Type": "AWS::ApiGateway::Deployment",
      "Properties": {
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        },
        "StageName": "dev"
      },
      "DependsOn": [
        "ApiGatewayMethodLambdaOptions",
        "ApiGatewayMethodLambdaGet"
      ]
    },
```

Added `__deployment__` to `serverless.yml`:

```yml
resources:
    __deployment__:
      DependsOn:
      - sample
```

Result:

```yml
    "ApiGatewayDeploymentXXXXXXXXX": {
      "Type": "AWS::ApiGateway::Deployment",
      "Properties": {
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        },
        "StageName": "dev"
      },
      "DependsOn": [
        "sample",
        "ApiGatewayMethodLambdaGet"
      ]
    },
```

Note how `sample` has replaced the first item in the list rather than being added to the list.

It would be great to be able to do the `$ApiGatewayDeployment` the way you described, but I'm concerned that the same merge issue would happen.

Is what I've described a bug in merging or expected behaviour? I would have expected this construct to replace the entire list (most predictable results) or to add to the list (either at the beginning or end, less predictable, no way to "empty out" the list with an override). Overwriting part of the list was not what I expected at all.

**Comment by medikoo** [comment]
> Note how sample has replaced the first item in the list rather than being added to the list.
> Is what I've described a bug in merging or expected behaviour?

 It's general design issue on framework side, which uses blind deep assign, and that's source of many issues. I've proposed a way how we can recover here https://github.com/serverless/serverless/issues/6575
I marked proposal as `breaking`, but technically support for `resources.extensions` could be implemented in non-breaking way.

**Comment by glb** [comment]
Thanks for the reference @medikoo! Would a PR that implements `resources.extensions` be accepted in a timely fashion? I'd be willing to give it a try, though I'm not certain that my JS skills are up to the challenge.

![image](https://user-images.githubusercontent.com/372794/74614262-7130a380-50e4-11ea-85fe-e67991f74de8.png)

I need to be able to have reliable deployments, and I'm not sure what other options there are.

If I understand correctly, you're suggesting that my use case would be solved by:

```yml
resources:
  extensions:
    __deployment__:
      DependsOn:
        - sample
```

(assuming I am using `serverless-plugin-bind-deployment-id`) or

```yml
resources:
  extensions:
    $ApiGatewayDeployment:
      DependsOn:
        - sample
```

in the hypothetical future where `$ApiGatewayDeployment` is a working alias for the generated deployment ID.

Yes? I assume that I'd start in `mergeCustomProviderResources.js`.

**Comment by medikoo** [comment]
> Would a PR that implements resources.extensions be accepted in a timely fashion? 

Definitely! I believe all implementation could be placed here: https://github.com/serverless/serverless/blob/3910df1ba6a8b39367ce8d51adb90216251be2ba/lib/plugins/aws/package/lib/mergeCustomProviderResources.js
and I think it shouldn't be difficult to accomplish (idea is quite simple)

> you're suggesting that my use case would be solved by (assuming I am using `serverless-plugin-bind-deployment-id`)

I think, it'll automatically pick up, yes.

> or in the hypothetical future where `$ApiGatewayDeployment` is a working alias for the generated deployment ID.

Yes, but I would treat it as separate enhancement to be addressed with different PR.

**Comment by glb** [comment]
OK, I will give it a try; if I can get something working I'll open a new PR.

**Comment by glb** [comment]
@medikoo you defined behaviour for `Properties` and `DependsOn` ... what should it do for other attributes? Just copy?

```
          switch (extensionAttributeName) {
            case 'Properties': ...
            case 'DependsOn': ...
            default: {
              template.Resources[resourceName][extensionAttributeName] = value;
            }
```

?

**Comment by medikoo** [comment]
> what should it do for other attributes? Just copy?

What are other known cases? I think we should either support them individually or throw with meaningful message that _extending `X` is not supported, please report if support is needed_

**Comment by glb** [comment]
There are [several](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-product-attribute-reference.html):

| Attribute name | Comments |
|---|---|
| CreationPolicy | object, looks like `_.merge` might be safe? |
| DeletionPolicy | simple string, simple overwrite? |
| Metadata | structure undefined, could be anything |
| UpdatePolicy | object, looks like `_.merge` might be safe? |
| UpdateReplacePolicy | simple string, simple overwrite? |
| _future_ | throw error as you suggested? |

Thoughts?

**Comment by medikoo** [comment]
-  `CreationPolicy`, `DeletionPolicy`, `UpdatePolicy`, `UpdateReplacePolicy`
I think we should overwrite. As what would be the use case to enrich configuration eventually setup by framework? It seems a bit quirky to me. I think in all cases full configuration should be defined in one place. So if user doesn't like what framework did, it's better if it overrides it completely.
- `Metadata`
I would merge with `Object.assign`
- _future_
`Sorry, extending of {attributeName} resource attribute at this point is not supported. Feel free to propose support for it in Framework issue tracker: https://github.com/serverless/serverless/issues`.

**Comment by glb** [comment]
OK. I think I've got the code, need to fill out the tests and write docs for this new behaviour and then will raise a PR, most likely tomorrow. 

Thank you!

**Comment by glb** [comment]
It looks like I'll need to make a PR to `serverless-plugin-bind-deployment-id` in order to get the behaviour that we're hoping for. 

jacob-meacham/serverless-plugin-bind-deployment-id#26

**Comment by glb** [comment]
Thanks for your help with #7352 @medikoo ! 

I haven't gotten any response to my issue + PR on `serverless-plugin-bind-deployment-id`, so it looks like I'll need to either create my own plugin or start down the path of creating an alias for the deployment resource ID like you mentioned [above](https://github.com/serverless/serverless/issues/7338#issuecomment-586754203). 

Any thoughts / suggestions on the best path?

**Comment by medikoo** [comment]
> or start down the path of creating an alias for the deployment resource ID like you mentioned above.
> Any thoughts / suggestions on the best path?

I think best would be to incorporate logic here: https://github.com/serverless/serverless/blob/08ec261a3cd34e7225f471cbeab8cef605ac61fc/lib/plugins/aws/package/lib/mergeCustomProviderResources.js#L31 (I would support it strictly in context of `resources.extensions`)

If `resourceName` name starts with an alias prefix (we should agree on something, I think `$` will work well). Then we assume it's an alias.

If it's a supported alias, replace it with resolved resource name, if not reject with meaningful error.

We can start just by supporting one `$ApiGatewayDeployment` alias, and eventually add others as demand will be reported.

@pmuens what are your thoughts?

**Comment by pmuens** [comment]
Hey @glb thanks for opening and @medikoo thanks for pinging.

We've introduced the concept of an `instanceId` to tackle exactly this kind of problem. Seems like it's quite hidden in the docs, but the initial PR for this feature can be found here --> https://github.com/serverless/serverless/pull/5926

The main idea is that the Serverless Framework generates a unique "instance ID" every time it runs. This instance id is accessible via Serverless Variables and used in all the places we rely on random numbers (i.e. in the case for a Deployment resource).

The docs can be found here --> https://serverless.com/framework/docs/providers/aws/guide/variables#referencing-serverless-core-variables

**Comment by glb** [comment]
Thanks @pmuens ... what I need to do is extend the `AWS::ApiGateway::Deployment` resource, which looks like this in the generated CloudFormation:

```json
    "ApiGatewayDeployment1111111111111": {
      "Type": "AWS::ApiGateway::Deployment",
      "Properties": {
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        },
        "StageName": "dev",
      },
      "DependsOn": [
        "ApiGatewayMethodLambdaOptions",
        "ApiGatewayMethodLambdaGet",
        "httpGet",
        "httpOptions"
      ]
    },
```

I was pretty certain that variable replacement didn't work in keys, but I tried it anyway:

```yaml
resources:
  extensions:
    ApiGatewayDeployment${sls:instanceId}:
      DependsOn:
        - sample
```

and I got:

```json
    "ApiGatewayDeployment${sls:instanceId}": {
      "DependsOn": [
        "sample"
      ]
    }
```

which is what I expected but not what I need.

@medikoo's suggestion of having `mergeCustomProviderResources.js` interpret a `$ApiGatewayDeployment` alias makes sense to me; it means I would write:

```yaml
resources:
  extensions:
    $ApiGatewayDeployment:
      DependsOn:
        - sample
```

and `sample` would be appended to the `DependsOn` list for the `AWS::ApiGateway::Deployment` resource (after the release that includes #7352 and this new change, of course).

Make sense?

**Comment by medikoo** [comment]
@pmuens it appears that in this case  relying on `sls:instanceId` variable won't solve the issue, as it's only property values and not names in which we resolve variables.

**Comment by pmuens** [comment]
I see, that makes sense 🤔

I'm fine with adding this functionality as it seems to be generic enough to be used in different contexts.

However we could also go another route and work on https://github.com/serverless/serverless/issues/2892 which is also a feature a lot of people have asked for and which can be used to resolve this problem without introducing a completely new concept.

**Comment by medikoo** [comment]
> However we could also go another route and work on #2892 which is also a feature a lot of people have asked for and which can be used to resolve this problem without introducing a completely new concept.

I totally agree, and I think it should not be that difficult to do (although already convoluted variables resolution logic may make it not that easy).

It'll definitely address broader range of eventual issues

**Comment by glb** [comment]
thanks @pmuens / @medikoo ... I'm not sure how to approach #2892. Thoughts?

**Comment by medikoo** [comment]
> thanks @pmuens / @medikoo ... I'm not sure how to approach #2892. Thoughts?

To provide good hint I'll need to deeply investigate current implementation (at this point I'm not fully familiar with it).

@erikerikson do you think it's easy to incorporate variable resolution on variable names? At least in certain data scopes?

**Comment by erikerikson** [comment]
Hi @medikoo - I've only just taken a quick look but a preliminary thought is that it should be possible while also a bit of careful, thoughtful work.  It would at least have to be done before context is added in getProperties and the multiple entry points (e.g. populateObject, populateService, prePopulateService) will likely require careful implementation.

**Comment by glb** [comment]
My PR for `serverless-plugin-bind-deployment-id` has been accepted (🎉) so I think @medikoo's suggestion of 

```yaml
resources:
  extensions:
    __deployment__:
      DependsOn:
        - sample
```

will work. Thanks everyone for their help!

**Comment by medikoo** [comment]
Great thanks @glb in helping to deliver `resources.extensions`! In my opinion it's a notable improvement to quality of a framework.
